## Market Basket Analysis (MBA)

A model that find combinations of items that occur together frequently in transactions and define relationships bewteen the items that people would buy.

## Concepts

**Assoication Rules** <br>
A model to analyze retail basket or transaction data and are intended to identify strong rules discovered in transaction data using measures of interestingness, based on the concept of strong rules.

$$ Rule = If \space an \space item \space A \space occurs, \space then \space item \space B \space also \space occurs \space with \space a \space certain \space probability.  $$
$$ X => Y, where \space X < I, Y < I \space and \space X \cup Y = 0 $$
<br>
<br>
**Apriori Algorithm** <br>
A classical algorithm in data mining, used for mining frequent itemsets and relevant association rules. It is devised to operate on a database containing a lot of transactions, for instance, items brought by customers in a store.

<Dataset>

**Item set**: A collection of one or more items.
$$ I = {i_1, i_2, ..., i_k} $$

**Transaction**: A record of a customer's purchase, containing one or more items.
$$ T = {t_1, t_2, ..., t_3} $$

In [ ]:
import pandas as pd

data_list = []
data = pd.DataFrame()

f = open("/content/groceries.csv")

for x in f:
  row = x.replace("\n","").split(",")
  data_list.append(row)

f.close()

for row in range(len(data_list)):
  for col in data_list[row]:
    data.loc[row,col] = 1

data.fillna(0, inplace=True)

<ipython-input-5-d99a09dc0877>:16: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  data.loc[row,col] = 1
<ipython-input-5-d99a09dc0877>:16: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  data.loc[row,col] = 1
<ipython-input-5-d99a09dc0877>:16: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  data.loc[row,c

In [ ]:
data

citrus fruit  semi-finished bread  margarine  ready soups  \
0              1.0                  1.0        1.0          1.0   
1              0.0                  0.0        0.0          0.0   
2              0.0                  0.0        0.0          0.0   
3              0.0                  0.0        0.0          0.0   
4              0.0                  0.0        0.0          0.0   
...            ...                  ...        ...          ...   
9830           1.0                  0.0        0.0          0.0   
9831           0.0                  0.0        0.0          0.0   
9832           1.0                  0.0        0.0          0.0   
9833           0.0                  1.0        0.0          0.0   
9834           0.0                  0.0        0.0          0.0   

      tropical fruit  yogurt  coffee  whole milk  pip fruit  cream cheese   \
0                0.0     0.0     0.0         0.0        0.0            0.0   
1                1.0     1.0     1.0         0.0        0.0            0.0   
2                0.0     0.0     0.0         1.0        0.0            0.0   
3                0.0     1.0     0.0         0.0        1.0            1.0   
4                0.0     0.0     0.0         1.0        0.0            0.0   
...              ...     ...     ...         ...        ...            ...   
9830             0.0     0.0     1.0         1.0        0.0            0.0   
9831             0.0     0.0     0.0         0.0        0.0            0.0   
9832             0.0     1.0     0.0         0.0        0.0            0.0   
9833             0.0     0.0     0.0         0.0        0.0            0.0   
9834             1.0     0.0     0.0         0.0        0.0            0.0   

      ...  baby food  pudding powder  tidbits  whisky  frozen fruits  bags  \
0     ...        0.0             0.0      0.0     0.0            0.0   0.0   
1     ...        0.0             0.0      0.0     0.0            0.0   0.0   
2     ...        0.0             0.0      0.0     0.0            0.0   0.0   
3     ...        0.0             0.0      0.0     0.0            0.0   0.0   
4     ...        0.0             0.0      0.0     0.0            0.0   0.0   
...   ...        ...             ...      ...     ...            ...   ...   
9830  ...        0.0             0.0      0.0     0.0            0.0   0.0   
9831  ...        0.0             0.0      0.0     0.0            0.0   0.0   
9832  ...        0.0             0.0      0.0     0.0            0.0   0.0   
9833  ...        0.0             0.0      0.0     0.0            0.0   0.0   
9834  ...        0.0             0.0      0.0     0.0            0.0   0.0   

      cooking chocolate  sound storage medium  kitchen utensil  \
0                   0.0                   0.0              0.0   
1                   0.0                   0.0              0.0   
2                   0.0                   0.0              0.0   
3                   0.0                   0.0              0.0   
4                   0.0                   0.0              0.0   
...                 ...                   ...              ...   
9830                0.0                   0.0              0.0   
9831                1.0                   0.0              0.0   
9832                0.0                   0.0              0.0   
9833                0.0                   0.0              0.0   
9834                0.0                   0.0              0.0   

      preservation products  
0                       0.0  
1                       0.0  
2                       0.0  
3                       0.0  
4                       0.0  
...                     ...  
9830                    0.0  
9831                    0.0  
9832                    0.0  
9833                    0.0  
9834                    0.0  

[9835 rows x 169 columns]

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9835 entries, 0 to 9834
Columns: 169 entries, citrus fruit to preservation products
dtypes: float64(169)
memory usage: 13.0 MB


In [ ]:
print(f"size {data.shape}")
print(f"Number of Transactions: {data.shape[0]}")
print(f"Number of Items: {data.shape[1]}")

size (9835, 169)
Number of Transactions: 9835
Number of Items: 169


## Metrics

In [ ]:
from itertools import combinations

all_items = data.columns.tolist()
item_pairs = list(combinations(all_items, 2))

In [ ]:
item_pairs[0:10]

[('citrus fruit', 'semi-finished bread'),
 ('citrus fruit', 'margarine'),
 ('citrus fruit', 'ready soups'),
 ('citrus fruit', 'tropical fruit'),
 ('citrus fruit', 'yogurt'),
 ('citrus fruit', 'coffee'),
 ('citrus fruit', 'whole milk'),
 ('citrus fruit', 'pip fruit'),
 ('citrus fruit', 'cream cheese '),
 ('citrus fruit', 'meat spreads')]

**Support** <br>
The number of transactions with both $X$ and $Y$ divided by the total number of transactions.
$$ supp(X => Y) = \frac{|X \cup Y|}{n} $$

In [ ]:
def support(X_union_Y_count, total_transaction):
  return X_union_Y_count / total_transactions

**Confidence** <br>
The percentage in which Y is bought with X.
$$ conf(x=>y) = \frac{supp(X \cup Y)}{supp(X)} $$

In [ ]:
def confidence(X_count, X_union_Y_count):
  return X_union_Y_count / X_count

**Lift** <br>
The ratio of the observed support to that expected if $X$ and $Y$ were independent.
$$ lift(X=>Y) = \frac{supp(X \cup Y)}{supp(X)supp(Y)}$$

In [ ]:
def lift(X_count, Y_count, X_union_Y_count, total_transaction):
  support_X = X_count / total_transactions
  support_Y = Y_count / total_transactions
  support_union = X_union_Y_count / total_transactions
  return support_union/(support_X * support_Y)

**Conviction** <br>
The ratio of the expected frequency that X occurs without Y X and Y were independent divided by the observed frequency of incorrect predictions.
$$ conv(X => Y) = \frac{1-supp(Y)}{1-conf(X=>Y)} $$

In [ ]:
def conviction(Y_count, total_transactions, confidence_value):
  support_Y = Y_count / total_transactions
  return (1-support_Y) / (1-confidence_value)

**Calculate Metrics**

1. $X$ = Number of transactions containing item X.
2. $Y$ = Number of transactions containing item Y.
3. $X \cup Y$ = Number of transactions containing both X and Y.
4. Total transactions = Total number of rows in the dataset.

In [ ]:
def calculate_metrics(data, X, Y):
    X_count = len(data[data[X].all(axis=1)])
    Y_count = len(data[data[Y].all(axis=1)])
    X_union_Y_count = len(data[(data[X] == 1).all(axis=1) & (data[Y] == 1).all(axis=1)])
    total_transactions = len(data)
    return X_count, Y_count, X_union_Y_count, total_transactions

Set the thresholds for some metrics (support and confidence).

In [ ]:
min_support = 0.01
min_confidence = 0.1

- `antecedent`: The item or set of items that appears before the 'then' in an association rule.
- `consequent`: The item or set of items that appears after the 'then' in an association rule.
- `support`: How popular an itemset is, as measured by the proportion of transactions in which an itemset appears. (How frequently a particular combination of items appears together in transactions.)
- `confidence`: The likelihood of item Y being purchased when item X is purchased, expressing the conditional probability of the consequent given the antecedent.
- `lift`: How much more likely item Y is purchased when item X is purchased, while controlling for how popular item Y is in general.
- `Conviction`: The probability that X appears without Y if they were independent with the actual frequency of incorrect predictions.

In [ ]:
results_df = pd.DataFrame(columns=["antecedent", "consequent", "support", "confidence", "lift", "conviction"])

In [ ]:
for X, Y in item_pairs:
    X_count, Y_count, X_union_Y_count, total_transactions = calculate_metrics(data, [X], [Y])

    support_value = support(X_union_Y_count, total_transactions)

    if support_value < min_support:
        continue

    confidence_value = confidence(X_count, X_union_Y_count)

    if confidence_value < min_confidence:
        continue

    lift_value = lift(X_count, Y_count, X_union_Y_count, total_transactions)
    conviction_value = conviction(Y_count, total_transactions, confidence_value)

    results_df = pd.concat([results_df, pd.DataFrame({
        "antecedent": [X],
        "consequent": [Y],
        "support": [support_value],
        "confidence": [confidence_value],
        "lift": [lift_value],
        "conviction": [conviction_value]
    })], ignore_index=True)

In [ ]:
results_df.sort_values(by='support', ascending=False)

antecedent          consequent   support  confidence  \
57              whole milk    other vegetables  0.074835    0.292877   
59              whole milk          rolls/buns  0.056634    0.221647   
36                  yogurt          whole milk  0.056024    0.401603   
66              whole milk     root vegetables  0.048907    0.191405   
85        other vegetables     root vegetables  0.047382    0.244877   
..                     ...                 ...       ...         ...   
115            frankfurter     root vegetables  0.010168    0.172414   
93                  butter  whipped/sour cream  0.010168    0.183486   
18               margarine                soda  0.010168    0.173611   
116            frankfurter             sausage  0.010066    0.170690   
124  fruit/vegetable juice             sausage  0.010066    0.139241   

         lift  conviction  
57   1.513634    1.140548  
59   1.205032    1.048452  
36   1.571735    1.244132  
66   1.756031    1.101913  
85   2.246605    1.179941  
..        ...         ...  
115  1.581800    1.076627  
93   2.559698    1.136928  
18   0.995607    0.999073  
116  1.816810    1.092534  
124  1.482068    1.052617  

[139 rows x 6 columns]

A support value of **0.07** for `{whole milk, other vegetables}` means that 7% of all transactions contain both whole milk and other vegetables.

In [ ]:
results_df.sort_values(by='confidence', ascending=False)

antecedent        consequent   support  confidence      lift  \
14          margarine        whole milk  0.024199    0.413194  1.617098   
21     tropical fruit        whole milk  0.042298    0.403101  1.577595   
36             yogurt        whole milk  0.056024    0.401603  1.571735   
2        citrus fruit        whole milk  0.030503    0.368550  1.442377   
4        citrus fruit  other vegetables  0.028876    0.348894  1.803140   
..                ...               ...       ...         ...       ...   
52             yogurt     domestic eggs  0.014337    0.102770  1.619775   
61         whole milk              curd  0.026131    0.102268  1.919481   
107     bottled water        newspapers  0.011286    0.102116  1.279376   
81   other vegetables              beef  0.019725    0.101944  1.943066   
32     tropical fruit       brown bread  0.010676    0.101744  1.568423   

     conviction  
14     1.268706  
21     1.247252  
36     1.244132  
2      1.179008  
4      1.238674  
..          ...  
52     1.043827  
61     1.054570  
107    1.024835  
81     1.055095  
32     1.041050  

[139 rows x 6 columns]

A confidence value of **0.413** for the rule `{margarine} -> {bread}` implies that 41.3% of the times a customer buys milk, they also buy bread.

In [ ]:
results_df.sort_values(by='lift', ascending=False)

antecedent          consequent   support  confidence      lift  \
113              beef     root vegetables  0.017387    0.331395  3.040367   
112              curd  whipped/sour cream  0.010473    0.196565  2.742150   
22     tropical fruit           pip fruit  0.020437    0.194767  2.574648   
93             butter  whipped/sour cream  0.010168    0.183486  2.559698   
122           chicken     root vegetables  0.010880    0.253555  2.326221   
..                ...                 ...       ...         ...       ...   
18          margarine                soda  0.010168    0.173611  0.995607   
119              soda     root vegetables  0.018607    0.106706  0.978964   
82   other vegetables                soda  0.032740    0.169207  0.970348   
62         whole milk                soda  0.040061    0.156785  0.899112   
7        citrus fruit                soda  0.012811    0.154791  0.887680   

     conviction  
113    1.332628  
112    1.155435  
22     1.147931  
93     1.136928  
122    1.193659  
..          ...  
18     0.999073  
119    0.997433  
82     0.993776  
62     0.979136  
7      0.976827  

[139 rows x 6 columns]

A lift value of **3.04** for the rule `{beef} -> {root vegetables}` indicates that customers who buy milk are 3.04 times more likely to buy bread compared to the average customer.

In [ ]:
results_df.sort_values(by='conviction', ascending=False)

antecedent        consequent   support  confidence      lift  \
113              beef   root vegetables  0.017387    0.331395  3.040367   
14          margarine        whole milk  0.024199    0.413194  1.617098   
21     tropical fruit        whole milk  0.042298    0.403101  1.577595   
36             yogurt        whole milk  0.056024    0.401603  1.571735   
4        citrus fruit  other vegetables  0.028876    0.348894  1.803140   
..                ...               ...       ...         ...       ...   
18          margarine              soda  0.010168    0.173611  0.995607   
119              soda   root vegetables  0.018607    0.106706  0.978964   
82   other vegetables              soda  0.032740    0.169207  0.970348   
62         whole milk              soda  0.040061    0.156785  0.899112   
7        citrus fruit              soda  0.012811    0.154791  0.887680   

     conviction  
113    1.332628  
14     1.268706  
21     1.247252  
36     1.244132  
4      1.238674  
..          ...  
18     0.999073  
119    0.997433  
82     0.993776  
62     0.979136  
7      0.976827  

[139 rows x 6 columns]

A conviction value of **1.33** for the rule `{beef} -> {root vegetables}` implies that the rule would be incorrect 1.3 times as often if the association between beef and root vegetables were purely random.